packages/yarnpkg-pnp/sources/esm-loader/loader.ts

```diff
@@ -0,0 +1,205 @@
+  try {
+    if (specifier.startsWith(`.`) && !(await isDirectory(parentPath))) {
+      parentPath = path.dirname(parentPath);
+    }
+  } catch {}
+
+  return new Promise((resolve, reject) => {
+    resolver.resolve({}, parentPath, specifier, {}, (err, file) => {
+      if (err || !file) {
+        reject(err);
+      } else {
+        resolve({url: pathToFileURL(file).href});
+      }
+    });
+  });
+}
+
+const realModules = new Set<string>();
+
+export async function getFormat(resolved: string, context: any, defaultGetFormat: any) {
+  const parsedURL = new URL(resolved);
+  if (parsedURL.protocol !== `file:`)
+    return defaultGetFormat(resolved, context, defaultGetFormat);
+
+  switch (path.extname(parsedURL.pathname)) {
+    case `.mjs`: {
+      realModules.add(fileURLToPath(resolved));
+      return {
+        format: `module`,
+      };
+    }
+    case `.json`: {
+      return {
+        format: `module`,
+      };
+    }
+    default: {
+      let packageJSONUrl = new URL(`./package.json`, resolved);
+      while (true) {
+        if (packageJSONUrl.pathname.endsWith(`node_modules/package.json`)) break;
+
+        const filePath = fileURLToPath(packageJSONUrl);
+
+        try {
+          let moduleType = (await readJson(filePath)).type ?? `commonjs`;
+          if (moduleType === `commonjs`)
+            moduleType = `module`;
+          else
+            realModules.add(fileURLToPath(resolved));
+
+          return {
+            format: moduleType,
+          };
+        } catch {}
+
+        const lastPackageJSONUrl = packageJSONUrl;
+        packageJSONUrl = new URL(`../package.json`, packageJSONUrl);
+
+        if (packageJSONUrl.pathname === lastPackageJSONUrl.pathname) {
+          break;
+        }
+      }
+    }
+  }
+
+  throw new Error(`Unable to get module type of '${resolved}'`);
+}
+
+let parserInit: Promise<void> | null = init().then(() => {
+  parserInit = null;
+});
+
+async function parseExports(filePath: string) {
+  const {exports} = parse(await readFile(filePath));
+
+  return new Set(exports);
+}
+
+export async function getSource(urlString: string, context: any, defaultGetSource: any) {
+  const url = new URL(urlString);
+  if (url.protocol !== `file:`) return defaultGetSource(url, context, defaultGetSource);
+
+  urlString = fileURLToPath(urlString);
+
+  if (realModules.has(urlString)) {
+    return {
+      source: await readFile(urlString),
+    };
+  }
+
+  if (parserInit !== null) await parserInit;
+
+  const exports = await parseExports(urlString);
+
+  let exportStrings = ``;
+  if (exports.has(`__esModule`)) {
+    for (const exportName of exports) {
+      if (exportName === `default`) {
+        exportStrings += `export default cjs['default']\n`;
+      } else {
+        exportStrings += `const __${exportName} = cjs['${exportName}'];\n export { __${exportName} as ${exportName} }\n`;
+      }
+    }
+  } else {
+    exportStrings = `export default cjs`;
+  }
```

Comment:
Node.js doesn't really care about `__esModule`

```suggestion
  for (const exportName of exports) {
    if (exportName === `default`) {
      exportStrings += `export default cjs['default']\n`;
    } else {
      exportStrings += `const __${exportName} = cjs['${exportName}'];\n export { __${exportName} as ${exportName} }\n`;
    }
  }
```

Reply:
I did that originally but it just crashed even harder than it does right now

Reply:
Actually, that might just be a bug in the babel code, https://github.com/babel/babel/blob/fed3fc06cf58949983d72746eee199d9e35c1565/packages%2Fbabel-types%2Fscripts%2Fgenerators%2Fast-types.js#L1 the file imported here does not have a default export so it should throw

Reply:
Mhh no. When importing a CJS file, the default export is _always_ present and it's set to `module.exports`.

Reply:
Maybe this?
```suggestion
  exportStrings += `export default cjs['default']\n`;

  for (const exportName of exports) {
    if (exportName !== `default`) {
      exportStrings += `const __${exportName} = cjs['${exportName}'];\n export { __${exportName} as ${exportName} }\n`;
    }
  }
```

Reply:
> Mhh no. When importing a CJS file, the default export is _always_ present and it's set to `module.exports`.

Ah, that was the missing information I needed for it to click, `yarn gulp` passes in that PR now, thanks :+1: 

> Maybe this?

That suggestion contradicts your previous comment

Reply:
> That suggestion contradicts your previous comment

Whoops, right. Well, your new commit looks correct :smile: